external/storm-kinesis/src/main/java/org/apache/storm/kinesis/spout/ExponentialBackoffRetrier.java

```diff
@@ -0,0 +1,164 @@
+    // Wait interfal for retrying after first failure
+    private final Long initialDelayMillis;
+    // Base for exponential function in seconds for retrying for second, third and so on failures
+    private final Long baseSeconds;
+    // Maximum number of retries
+    private final Long maxRetries;
+    // map to track number of failures for each kinesis message that failed
+    private Map<KinesisMessageId, Long> failCounts = new HashMap<>();
+    // map to track next retry time for each kinesis message that failed
+    private Map<KinesisMessageId, Long> retryTimes = new HashMap<>();
+    // sorted set of records to be retrued based on retry time. earliest retryTime record comes first
+    private SortedSet<KinesisMessageId> retryMessageSet = new TreeSet<>(new RetryTimeComparator());
+
+    /**
+     * no args constructor that uses defaults of 100 ms for first retry, max retries of Long.MAX_VALUE and an exponential backoff of Math.pow(2,i-1) secs for
+     * retry i where i = 2,3,
+     */
+    public ExponentialBackoffRetrier () {
+        this(100L, 2L, Long.MAX_VALUE);
+    }
+
+    /**
+     *
+     * @param initialDelayMillis delay in milliseconds for first retry
+     * @param baseSeconds base for exponent function in seconds
+     * @param maxRetries maximum number of retries before the record is discarded/acked
+     */
+    public ExponentialBackoffRetrier (Long initialDelayMillis, Long baseSeconds, Long maxRetries) {
+        this.initialDelayMillis = initialDelayMillis;
+        this.baseSeconds = baseSeconds;
+        this.maxRetries = maxRetries;
+        validate();
+    }
+
+    private void validate () {
+        if (initialDelayMillis < 0) {
+            throw new IllegalArgumentException("initialDelayMillis cannot be negative." );
+        }
+        if (baseSeconds < 0) {
+            throw new IllegalArgumentException("baseSeconds cannot be negative.");
+        }
+        if (maxRetries < 0) {
+            throw new IllegalArgumentException("maxRetries cannot be negative.");
+        }
+    }
+    @Override
+    public boolean failed(KinesisMessageId messageId) {
+        LOG.debug("Handling failed message " + messageId);
+        // if maxRetries is 0, dont retry and return false as per interface contract
+        if (maxRetries == 0) {
+            LOG.debug("maxRetries set to 0. Hence not queueing " + messageId);
+            return false;
+        }
+        // if first failure add it to the count map
+        if (!failCounts.containsKey(messageId)) {
+            failCounts.put(messageId, 0L);
+        }
+        // increment the fail count as we started with 0
+        Long failCount = failCounts.get(messageId);
+        failCounts.put(messageId, ++failCount);
+        // if fail count is greater than maxRetries, discard or ack. for e.g. for maxRetries 3, 4 failures are allowed at maximum
+        if (failCount > maxRetries) {
+            LOG.debug("maxRetries reached so dropping " + messageId);
```

Comment:
can you make this as warn as it can be useful to see this in the logs by default